### ISSUE TITLE
Carry triggerinstance context everywhere

### ISSUE THREAD

**Issue by manasdk** [StackStorm/st2#1602]
1. add triggerinstance.correlation_id to supply reference to an external event in a triggerinstance
2. triggerinstance.id accessible in a rule so that it can be passed into an action
2. `on-start` notification to notify when an execution is started
2. `ActionTrigger_start` TriggerType
2. `triggerinstance.id` and `triggerinstance.correlation_id` should be available in notifier, actiontrigger etc.

**Comment by skarmark** [comment]
Little bit of a context of the use case which we talked about -

Recently I added the ability to be able to filter executions by trigger instance id. This is useful when you know the trigger instance id or can look for it easily. For thousands of executions run in response to sensors listening to messaging queue such as rabbitmq, it is hard to find a particular execution. There a couple of ways to solve this - 

1. Give ability for users to create their own IDs eg. UUID inside the sensor and pass it as correlation_id when dispatching trigger (#1 and maybe #5 above).
2. Make triggerinstance.id accessible in the sensor or rule so that it can be communicated to the user (#  2, 3 and 4 above).

Personally, I like option 1 better. This way, stackstorm doesn't need to communicate the id, since user already knows it. And this being an attribute of trigger instance, it makes it easy to filter by using stackstorm api or cli or even webui. If we use a custom attribute in payload to store this, users will need to explicitly pass it through all the actions in the workflow and filtering is harder as well. I believe this should be pretty easy to add as well.
